### PR TITLE
Ensure answers provided to SingleListGrader are nonempty

### PR DESCRIPTION
**Pull request by jolyonb**
This validates that the configuration for `SingleListGrader` does not have any empty answers, if such answers would require the student to enter an answer that would run afoul of the `missing_error` setting.

### CODE REVIEW

File: mitxgraders/listgrader.py [hunk 1]
```diff
@@ -682,6 +683,12 @@ def schema_answers(self, answer_tuple):
             if self.config['delimiter'] == subgrader.config['delimiter']:
                 raise ConfigError("Nested SingleListGraders must use different delimiters.")
 
+        # Check for empty entries anywhere in answers_tuple (which can be a nested mess!)
+        # We do this before validating answer_list, as strings may be coerced into other
+        # objects by schema validation (e.g., FormulaGrader coerces expect into a dict)
```
Comment (ChristopherChudzicki) [inline mitxgraders/listgrader.py]: 👍

File: mitxgraders/listgrader.py [hunk 2]
```diff
@@ -734,3 +741,33 @@ def check_response(self, answer, student_input, **kwargs):
...
+            demand_no_nonempty(item)
+    else:
+        test = obj
+        if isinstance(test, dict):
+            if 'expect' in test:
+                test = test['expect']
+            else:
+                return
+        # We're doing this BEFORE validation, so could be given text or a string
+        if isinstance(test, six.text_type) or isinstance(test, six.string_types):
```
Comment (ChristopherChudzicki) [inline mitxgraders/listgrader.py]: You only need `isinstance(test, six.string_types)` here ( `(basestring,)` in 2, `(str,)` in 3)

File: mitxgraders/listgrader.py [hunk 3]
```diff
@@ -734,3 +741,33 @@ def check_response(self, answer, student_input, **kwargs):
                                         partial_credit=self.config['partial_credit'])
 
         return result
+
+def demand_no_nonempty(obj):
```
Comment (ChristopherChudzicki) [inline mitxgraders/listgrader.py]: should this be `demand_all_nonempty`? or `demand_no_empty`?

File: mitxgraders/listgrader.py (hunk 2 above)
Comment (jolyonb) [reply mitxgraders/listgrader.py]: Happy to make the change, but if someone imports unicode literals, then only doing `string_types` will fail though, right?

File: mitxgraders/listgrader.py (hunk 3 above)
Comment (jolyonb) [reply mitxgraders/listgrader.py]: Oops! Fixed.

File: mitxgraders/listgrader.py (hunk 2 above)
Comment (ChristopherChudzicki) [reply mitxgraders/listgrader.py]: Should work... `basestring` is the superclass of string and unicode.

```python
import six

# Python 2.7
isinstance(u'cat', six.string_types) # True
isinstance('cat', six.string_types) # True
```

Comment (jolyonb) [reply mitxgraders/listgrader.py]: Good point. Now fixed.

**Comment by jolyonb** [comment]
Rebase coming up...

**Pull request merged**